server/schedule/operator/builder.go

```diff
@@ -183,99 +235,190 @@ func (b *Builder) SetLightWeight() *Builder {
+	b.toPromote = newPeersMap()
+	b.toDemote = newPeersMap()
+
+	voterCount := 0
+	for _, peer := range b.targetPeers {
+		if !core.IsLearner(peer) {
 			voterCount++
 		}
 	}
 	if voterCount == 0 {
 		return "", errors.New("cannot create operator: target peers have no voter")
 	}
 
-	// Diff `originPeers` and `targetPeers` to initialize `toAdd`,
-	// `toPromote`, `toRemove`.
-	for _, o := range b.originPeers.m {
-		n := b.targetPeers.Get(o.GetStoreId())
-		// no peer in targets, or target is learner while old one is voter.
-		if n == nil || (core.IsLearner(n) && !core.IsLearner(o)) {
+	// Diff `originPeers` and `targetPeers` to initialize `toAdd`, `toRemove`, `toPromote`, `toDemote`.
+	// Note: Use `toDemote` only when `useJointConsensus` is true. Otherwise use `toAdd`, `toRemove` instead.
+	for _, o := range b.originPeers {
+		n := b.targetPeers[o.GetStoreId()]
+		if n == nil {
 			b.toRemove.Set(o)
 			continue
 		}
-		if core.IsLearner(o) && !core.IsLearner(n) {
-			b.toPromote.Set(n)
+
+		// If the peer id in the target is different from that in the origin,
+		// modify it to the peer id of the origin.
+		if o.GetId() != n.GetId() {
+			n = &metapb.Peer{
+				Id:      o.GetId(),
+				StoreId: o.GetStoreId(),
+				Role:    n.GetRole(),
+			}
+		}
+
+		if core.IsLearner(o) {
+			if !core.IsLearner(n) {
+				// learner -> voter
+				b.toPromote.Set(n)
+			}
+		} else {
+			if core.IsLearner(n) {
+				// voter -> learner
+				if b.useJointConsensus {
+					b.toDemote.Set(n)
+				} else {
+					b.toRemove.Set(o)
+					// Need to add `b.toAdd.Set(n)` in the later targetPeers loop
+				}
+			}
 		}
 	}
-	for _, n := range b.targetPeers.m {
-		o := b.originPeers.Get(n.GetStoreId())
-		if o == nil || (core.IsLearner(n) && !core.IsLearner(o)) {
-			// old peer not exists, or target is learner while old one is voter.
+	for _, n := range b.targetPeers {
+		// old peer not exists, or target is learner while old one is voter.
+		o := b.originPeers[n.GetStoreId()]
+		if o == nil || (!b.useJointConsensus && !core.IsLearner(o) && core.IsLearner(n)) {
 			if n.GetId() == 0 {
 				// Allocate peer ID if need.
 				id, err := b.cluster.AllocID()
 				if err != nil {
 					return "", err
 				}
-				n.Id = id
+				n = &metapb.Peer{
+					Id:      id,
+					StoreId: n.GetStoreId(),
+					Role:    n.GetRole(),
+				}
 			}
+			// It is a pair with `b.toRemove.Set(o)` when `o != nil`.
 			b.toAdd.Set(n)
 		}
 	}
 
-	b.currentPeers, b.currentLeader = b.originPeers.Copy(), b.originLeader
+	// If the target leader does not exist or is a Learner, the target is cancelled.
+	if peer, ok := b.targetPeers[b.targetLeaderStoreID]; !ok || core.IsLearner(peer) {
+		b.targetLeaderStoreID = 0
+	}
+
+	// If no target leader is specified, try not to change the leader as much as possible.
+	if b.targetLeaderStoreID == 0 {
+		if peer, ok := b.targetPeers[b.originLeaderStoreID]; ok && !core.IsLearner(peer) {
+			b.targetLeaderStoreID = b.originLeaderStoreID
+		}
+	}
```

Comment:
Do we need to add a check `if originLeaderStoreID == 0 ` and throw the error if it does?

Reply:
Before that, it would have an error in `NewBuilder`.